pageserver/src/basebackup.rs

```diff
@@ -228,9 +228,14 @@ impl<'a> Basebackup<'a> {
         pg_control.state = pg_constants::DB_SHUTDOWNED;
 
         // add zenith.signal file
+        let xl_prev = if self.prev_record_lsn == Lsn(0) {
+            checkpoint.redo - 8 // subtract 8 to preserve alignment on WAL record boindary
+        } else {
+            self.prev_record_lsn.0
+        };
```

Comment:
I don't understand this. How does subtracting 8 help with alignment? Surely `checkpoint.redo` is 8-bytes aligned to begin with, and subtracting 8 doesn't change that.

Why is `checkpoint.redo-8` the correct value for xl_prev in this case?

If the page server doesn't know what the correct xl_prev value would be, I think '0' is a good way to indicate that. The code in postgres startup can do what it needs to, to start up without a valid xl_prev value.

Reply:
> Why is checkpoint.redo-8 the correct value for xl_prev in this case?

I think idea was to have any LSN that is less then checkpoint.redo, and not to break alignment by subtracting 1.

But that approach is bad because we would not know after was that lsn valid in the first place or not. So zero or some constant like 0xBAD seems to be a better option.

Reply:
Sorry, that my comment was not clear.
We need to enforce xl_prev<curr_lsn condition (checked by XLogReader).
Subtracting 8 doesn't enforce alignment, it preserves alignment.
0 is considered as invalid LSN and rejected by XLogReader